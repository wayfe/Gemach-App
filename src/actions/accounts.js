import database from "../firebase/firebase";


// add account redux
export const addAccount = (account) => ({
    type: "ADD_ACCOUNT",
    account,
});

// add account firebase
export const startAddAccount = (accountData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            name = "", 
            phoneNum= "", 
            address = "", 
            address2 = "",
            city = "",
            state = "",
            zip = "",
            id = ""
        } = accountData;
        const account = { name, phoneNum, address, address2, city, state, zip, transactions: [] }
        return database.ref(`/users/${uid}/accounts/${id}`).set(account).then(() => {
            dispatch(addAccount({
                id,
                ...account
            }));
        });
    };                                                                                                                       
};

// remove account
export const removeAccount = ({ id } = {}) => ({
    type: "REMOVE_ACCOUNT",
    id
});

// edit account
export const editAccount = (id, updates) => ({
    type: "EDIT_ACCOUNT",
    id,
    updates
});

// edit account firebase
export const startEditAccount = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${id}`).update(updates).then(() => {
            dispatch(editAccount(id, updates));
        })
    }
}

// add transactions
export const addTransaction = (id, transaction) => ({
    type: "ADD_TRANSACTION",
    id,
    transaction
});


export const startAddTransaction = (accountId, transaction = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            id = "",
            amount = "",
            note = "",
            date = "",
            type = "",
            payType = "",
            checkNo = "",
            isCleared = false,
        } = transaction
        const transactionData = { amount, note, date, type, payType, isCleared, checkNo}
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${id}`).set(transactionData)
        .then(() => {
            dispatch(addTransaction(accountId, transaction));
        });
    };
};

// remove transactions
export const removeTransaction = (accountId, transactionId) => ({
    type: "REMOVE_TRANSACTION",
    accountId,
    transactionId
});

export const startRemoveTransaction = (accountId, transactionId) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${transactionId}`).set(null).then(() => {
            dispatch(removeTransaction(accountId, transactionId));
        });
    }
}

// edit transactions
export const editTransaction = (accountId, transaction) => ({
    type: "EDIT_TRANSACTION",
    accountId,
    transaction
})

export const startEditTransaction = (accountId, transaction) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${transaction.id}`).update(transaction).then(() => {
            dispatch(editTransaction(accountId, transaction));
        })
    }
}

// SET_ACCOUNTS
export const setAccounts = (accounts) => ({
    type: "SET_ACCOUNTS",
    accounts
});


export const startSetAccounts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/`).once("value").then((snapshots) => {
            const accounts = [];
            snapshots.forEach((i) => {
                const transactionArray = [] 
                const transactions = i.val().transactions;

                for (var key in transactions) {
                    transactionArray.push({...transactions[key], id: key});
                }
                    
                const account = {...i.val(), id: i.key, transactions: transactionArray}
                    
                accounts.push(account);
            });
            console.log(accounts);
            dispatch(setAccounts(accounts));
        });
    };                                                                                                                       
};
