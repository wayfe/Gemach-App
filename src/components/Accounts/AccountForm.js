import React from "react";

export default class AccountForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            name: props.account ? props.account.name : "",
            phoneNum: props.account ? props.account.phoneNum : "",
            address: props.account ? props.account.address : "",
            address2: props.account ? props.account.address2 : "",
            city: props.account ? props.account.city : "",
            state: props.account ? props.account.state : "",
            zip: props.account ? props.account.zip : "",
            error: []
        }
    }

    onNameChange = (e) => {
        const re = /^([^0-9]*)$/;

        const inputValue = e.target.value;

        const inputName = e.target.name;

        if (!inputValue || re.test(inputValue)) {
            this.setState(() => ({[inputName] : inputValue}));
        }

    }

    onInputChange = (e) => {
        const name = e.target.name;
        const value = e.target.value;
        this.setState(() => ({[name] : value}));
    }

    onNumChange = (e) => {

        const re = /^[0-9\b]+$/;

        const name = e.target.name;
        
        const num = e.target.value;

        if (!num || re.test(num)) {
            this.setState(() => ({ [name] : num }));
        }
    }

    onSubmit = (e) => {
        e.preventDefault();

        const phoneRegex = /^\D?(\d{3})\D?\D?(\d{3})\D?(\d{4})$/;

        const error = [];
        
        ["name", "phoneNum", "address", "city", "state", "zip"].forEach((input) => {
            if (input === "phoneNum" && !phoneRegex.test(this.state.phoneNum)) {
                error.push(input)
            } else if (!this.state[input]) {
                error.push(input)
            }
        });

        if (error.length > 0) {
            this.setState(() => ({error : error, errorMessage: "hmmm, something looks wrong here..."}));
        } else {
            this.props.onSubmit({
                name: this.state.name,
                phoneNum: this.state.phoneNum,
                address: this.state.address,
                address2: this.state.address2,
                city: this.state.city,
                state: this.state.state,
                zip: this.state.zip
            });
        }
    }

    render () {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.error.includes("name") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="name"
                        name="name"
                        autoFocus
                        value={this.state.name}
                        onChange={this.onNameChange}
                    />

                    {this.state.error.includes("phoneNum") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="phone"
                        name="phoneNum"
                        value={this.state.phoneNum}
                        onChange={this.onNumChange}
                    />

                    {this.state.error.includes("address") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="address"
                        name="address"
                        value={this.state.address}
                        onChange={this.onInputChange}
                    />
                    <input 
                        type="text"
                        placeholder="address line 2"
                        name="address2"
                        value={this.state.address2}
                        onChange={this.onInputChange}
                    />

                    {this.state.error.includes("city") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="city"
                        name="city"
                        value={this.state.city}
                        onChange={this.onNameChange}
                    />

                    {this.state.error.includes("state") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="state"
                        name="state"
                        value={this.state.state}
                        onChange={this.onNameChange}
                    />

                    {this.state.error.includes("zip") && <p>{this.state.errorMessage}</p>}
                    <input 
                        type="text"
                        placeholder="zip"
                        name="zip"
                        value={this.state.zip}
                        onChange={this.onNumChange}
                    />
                    <button>Save</button>
                </form>
            </div>
        );
    }
}


