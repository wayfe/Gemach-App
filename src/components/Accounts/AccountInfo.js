import React from "react";

const AccountInfo = ({account}) => (
    <div>
        <p>{account.name}</p>
        <p>{account.phoneNum}</p>
        <p>{account.address}</p>
        {account.address2 && <p>{account.address2}</p>}
        <p>{account.city}</p>
        <p>{account.state}</p>
        <p>{account.zip}</p>
    </div>
)

export default AccountInfo;