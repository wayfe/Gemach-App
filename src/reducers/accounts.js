
const AccountsReducerDefaultState = [];

export default (state = AccountsReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_ACCOUNT":
            return [
                ...state, action.account
            ];
        case "REMOVE_ACCOUNT":
            return state.filter(({ id }) => id !== action.id);
        case "EDIT_ACCOUNT":
            return state.map((account) => {
                if (account.id === action.id) {
                    return {
                        ...account, ...action.updates
                    };
                } else {
                    return account
                }
            });
        case "ADD_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.id) {
                return {
                    ...account, transactions: [...account.transactions, action.transaction]
                };
            } else {
                return account
            }
        });
        case "REMOVE_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.accountId) {
                return {
                    ...account, transactions: account.transactions.filter((transaction) => transaction.id !== action.transactionId)
                };
            } else {
                return account
            }
        });
        case "EDIT_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.accountId) {
                return {
                    ...account, transactions: [...account.transactions.filter((transaction) => transaction.id !== action.transaction.id), action.transaction]
                };
            } else {
                return account
            }
        });
        case "SET_ACCOUNTS":
            return action.accounts;
        default:
            return state;
    }
};
