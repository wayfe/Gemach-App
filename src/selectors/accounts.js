
//Get Visible Data
export default (accounts, {text}) => {
    return accounts.filter((account) => {
        const textMatch = account.name.toLowerCase().includes(text.toLowerCase());

        return textMatch;
    }).sort((a,b) => {return a.name > b.name ? 1 : -1})
}

