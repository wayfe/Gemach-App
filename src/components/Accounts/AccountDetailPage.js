import React from "react";
import {startEditAccount} from "../../actions/accounts";
import { connect } from "react-redux";
import {withRouter} from "react-router-dom";
import TransactionsList from "../Transactions/TransactionsList";
import AccountInfo from "./AccountInfo";
import AccountForm from "./AccountForm";
import DataListFilters from "../Filters/DataListFilters";
import AddTransactionPage from "../Transactions/AddTransactionsPage";
import Balance from "../Balance";

class AccountDetailPage extends React.Component {

    constructor(props) {
        super(props)

        this.state = {
            clicked: false
        }
    }

    onEditAccount = () => {
        this.setState(() => ({
            clicked: true
        }))
    }

    onSubmit = (account) => {
        this.props.dispatch(startEditAccount(this.props.account.id, account));
        this.setState(() => ({
            clicked: false
        }))
    }

    render() {
        return (
            <div>
                {this.state.clicked === false ?
                    <div>
                        <AccountInfo 
                            account={this.props.account} 
                        />
                        <button
                            onClick={this.onEditAccount}
                        >edit account info</button>
                    </div> 
                    :
                    <AccountForm 
                        account={this.props.account}
                        onSubmit={this.onSubmit}
                    />
                }
                <Balance 
                    transactions={this.props.account.transactions}
                />
                <AddTransactionPage 
                    account={this.props.account}
                />
                {this.props.account.transactions.length > 0 &&
                    <div>
                        <DataListFilters 
                            activeFilters={["transactionFilters"]}
                        /> 
                        <TransactionsList 
                            transactions={this.props.account.transactions}
                        />
                    </div> 
                } 
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        account: state.accounts.find((account) => account.id === props.match.params.id)
    };
};

export default withRouter(connect(mapStateToProps)(AccountDetailPage));


