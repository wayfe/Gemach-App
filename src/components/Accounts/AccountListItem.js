import React from "react";
import {Link} from "react-router-dom";
import {getBal} from "../../selectors/balance";
import {connect} from "react-redux";
import numeral from 'numeral';

export const AccountListItem = (props) => (
    <div>
        <Link to={`/accounts/${props.account.id}`}><h3>{props.account.name}</h3></Link>
        {props.account.transactions.length > 0 && <p>Balance: {numeral(getBal(props.account.transactions) /100).format('$0,0.00')}</p>}
    </div>
);

export default AccountListItem;
