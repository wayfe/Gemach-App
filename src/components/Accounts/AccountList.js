import React from "react";
import {connect} from "react-redux";
import AccountListItem from "./AccountListItem";
import selectAccounts from "../../selectors/accounts";

export const AccountList = (props) => (
    <div>
        {props.accounts.length < 1 ? <p>No Accounts</p> : props.accounts.map((account) => 
            <AccountListItem key={account.id} account={account} />
        )}
    </div>
);

const mapStateToProps = (state) => {
    return {
        accounts: selectAccounts(state.accounts, state.filters)
    }
};

export default connect(mapStateToProps)(AccountList);



