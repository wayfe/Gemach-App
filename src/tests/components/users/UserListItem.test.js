import React from "react";
import {shallow} from "enzyme";
import { AccountListItem } from "../../../components/Accounts/AccountListItem";
import accounts from "../../fixtures/accounts";

test("should render AccountListItem for a account", () => {
    const wrapper = shallow(<AccountListItem account={accounts[0]} />);
    expect(wrapper).toMatchSnapshot();
})

