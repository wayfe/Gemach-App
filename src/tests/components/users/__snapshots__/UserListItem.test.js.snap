// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AccountListItem for a account 1`] = `
<div>
  <Link
    replace={false}
    to="/accounts/1"
  >
    <h3>
      Katz
    </h3>
  </Link>
  <p>
    Balance: 
    $220.00
  </p>
</div>
`;
